File: src/0784.letter-case-permutation.800/solution.ts
function letterCasePermutation (S: string): string[] {
    const result:string[] = [];
    backTracking(S, [] as string[], result);
    return result;
}

function backTracking (S:string, sequence:string[], result:string[]) {
    if (sequence.length === S.length) {
        result.push(sequence.join(''));
        return;
    }
    const code = S.charCodeAt(sequence.length);
    if (code > 47 && code < 58) {
        sequence.push(S[sequence.length]);
        backTracking(S, sequence, result);
        sequence.pop();
    } else {
        sequence.push(S[sequence.length].toLowerCase());
        backTracking(S, sequence, result);
        sequence.pop();

        sequence.push(S[sequence.length].toUpperCase());
        backTracking(S, sequence, result);
        sequence.pop();
    }
}

File: src/0784.letter-case-permutation.800/solution.java
class Solution {
    char[] chars;
    List<String> result;
    public List<String> letterCasePermutation(String s) {
        chars = s.toCharArray();
        result = new ArrayList<>();
        backTracking(0,new StringBuilder());
        return result;
    }

    private void backTracking(int index,StringBuilder stringBuilder){
        if(index == chars.length){
            result.add(stringBuilder.toString());
            return;
        }
        if(Character.isDigit(chars[index])){
            stringBuilder.append(chars[index]);
            backTracking(index+1,stringBuilder);
            stringBuilder.deleteCharAt(stringBuilder.length()-1);
        }else{
            stringBuilder.append(Character.toUpperCase(chars[index]));
            backTracking(index+1,stringBuilder);
            stringBuilder.deleteCharAt(stringBuilder.length()-1);

            stringBuilder.append(Character.toLowerCase(chars[index]));
            backTracking(index+1,stringBuilder);
            stringBuilder.deleteCharAt(stringBuilder.length()-1);
        }


    }

}